Run PreStmt/PostStmt checker for GCCAsmStmt

T-Gruber opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/94940

Run PreStmt and PostStmt checker for GCCAsmStmt.
Unittest to validate that corresponding callback functions are triggered.

T-Gruber committed:
Run PreStmt/PostStmt checker for GCCAsmStmt

steakhal commented:
Could you please drop the format only changes so that I could focus on the actual change set?

T-Gruber committed:
Restored format of previous commit

T-Gruber committed:
Removed unused locals in code snippets

T-Gruber commented:
> Could you please drop the format only changes so that I could focus on the actual change set?

I just reverted the format changes. Thanks for letting me know.

T-Gruber commented:
Hi @steakhal , what is your opinion on this PR?
Are there any open points from my side?

steakhal commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -2057,11 +2057,16 @@ void ExprEngine::Visit(const Stmt *S, ExplodedNode *Pred,
+      getCheckerManager().runCheckersForPostStmt(Dst, PostVisit, S, *this);
       Bldr.addNodes(Dst);
-      break;
+    } break;
```
Hm, I think it would look better if the break was inside the braces.

steakhal commented on clang/unittests/StaticAnalyzer/ExprEngineVisitTest.cpp:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+class ExprEngineVisitPreChecker : public Checker<check::PreStmt<GCCAsmStmt>> {
+public:
+  void checkPreStmt(const GCCAsmStmt *ASM, CheckerContext &C) const {
+    emitErrorReport(C, Bug, "PreStmt<GCCAsmStmt>");
+  }
+
+private:
+  const BugType Bug{this, "GCCAsmStmtBug"};
+};
+
+class ExprEngineVisitPostChecker : public Checker<check::PostStmt<GCCAsmStmt>> {
+public:
+  void checkPostStmt(const GCCAsmStmt *ASM, CheckerContext &C) const {
+    emitErrorReport(C, Bug, "PostStmt<GCCAsmStmt>");
+  }
+
+private:
+  const BugType Bug{this, "GCCAsmStmtBug"};
+};
```
I think you could merge these two into a single class.
Similar goes to the other parts. To me `asm stmt` is itself a thing to test.
But it's just a personal preference.
One other way to look at this is basically the combinatorical explosion of the tests if we wanted to cover all other statements and different callback kinds. The boilerplate to useful code would just blow up.

steakhal reviewed (changes requested):
Thanks for the ping. Looks really nice!
FYI I just came across a case last week w.r.t asm gotos. You can have a read [here](https://sonarsource.atlassian.net/browse/CPP-5459) if you are interested in the subject.

T-Gruber committed:
Move break into braces

T-Gruber committed:
Adapted format of file header

T-Gruber commented:
> Thanks for the ping. Looks really nice! FYI I just came across a case last week w.r.t asm gotos. You can have a read [here](https://sonarsource.atlassian.net/browse/CPP-5459) if you are interested in the subject.

Thank you! Looks interesting. Does this mean that the linked problem would also benefit from this PR?

steakhal commented:
> > Thanks for the ping. Looks really nice! FYI I just came across a case last week w.r.t asm gotos. You can have a read [here](https://sonarsource.atlassian.net/browse/CPP-5459) if you are interested in the subject.
> 
> Thank you! Looks interesting. Does this mean that the linked problem would also benefit from this PR?

I haven't if this PR would help or not. My guess would be that this wouldn't help. 

Please fix the formatting. After that we can merge this PR.

T-Gruber committed:
Reworked unittest

T-Gruber commented:
> > > Thanks for the ping. Looks really nice! FYI I just came across a case last week w.r.t asm gotos. You can have a read [here](https://sonarsource.atlassian.net/browse/CPP-5459) if you are interested in the subject.
> > 
> > 
> > Thank you! Looks interesting. Does this mean that the linked problem would also benefit from this PR?
> 
> I haven't if this PR would help or not. My guess would be that this wouldn't help.
> 
> Please fix the [formatting](https://github.com/llvm/llvm-project/pull/95409#issuecomment-2219921933). After that we can merge this PR.

I reworked the unittest and fixed the format as well. Thanks for your help!
